native/src/widget/text_input.rs

```diff
@@ -1041,3 +841,182 @@ mod platform {
+    }
+
+    impl Default for Cursor {
+        fn default() -> Self {
+            Cursor {
+                start: 0,
+                end: 0,
+                click_count: 0,
+                last_click_position: None,
+                last_click_timestamp: None,
+            }
+        }
+    }
+
+    impl Cursor {
+        /* Move section */
+        pub fn move_to(&mut self, position: usize) {
+            self.start = position;
+            self.end = position;
+        }
+
+        pub fn move_right(&mut self, value: &Value) {
+            if self.is_selection() {
+                let dest = self.right();
+                self.start = dest;
+                self.end = dest;
+            } else if self.end < value.len() {
+                self.start += 1;
+                self.end += 1;
+            }
+        }
+
+        pub fn move_left(&mut self) {
+            if self.is_selection() {
+                let dest = self.left();
+                self.start = dest;
+                self.end = dest;
+            } else if self.left() > 0 {
+                self.start -= 1;
+                self.end -= 1;
+            }
+        }
+
+        pub fn move_right_by_amount(&mut self, value: &Value, amount: usize) {
+            self.start = self.start.saturating_add(amount).min(value.len());
+            self.end = self.end.saturating_add(amount).min(value.len());
+        }
+
+        pub fn move_left_by_words(&mut self, value: &Value) {
+            let (left, _) = self.cursor_position(value);
+
+            self.move_to(value.previous_start_of_word(left));
+        }
+
+        pub fn move_right_by_words(&mut self, value: &Value) {
+            let (_, right) = self.cursor_position(value);
+
+            self.move_to(value.next_end_of_word(right));
+        }
+        /* Move section end */
+
+        /* Selection section */
+        pub fn select_range(&mut self, start: usize, end: usize) {
+            self.start = start;
+            self.end = end;
+        }
+
+        pub fn select_left(&mut self) {
+            if self.end > 0 {
+                self.end -= 1;
+            }
+        }
+
+        pub fn select_right(&mut self, value: &Value) {
+            if self.end < value.len() {
+                self.end += 1;
+            }
+        }
+
+        pub fn select_left_by_words(&mut self, value: &Value) {
+            self.end = value.previous_start_of_word(self.start);
+        }
+
+        pub fn select_right_by_words(&mut self, value: &Value) {
+            self.end = value.next_end_of_word(self.start);
+        }
+
+        pub fn select_all(&mut self, len: usize) {
+            self.start = 0;
+            self.end = len;
+        }
+        /* Selection section end */
+
+        /* Double/Triple click section */
+        // returns the amount of clicks on the same position in specific timeframe
+        // (1=double click, 2=triple click)
+        pub fn process_click(&mut self, position: Point) -> usize {
```

Comment:
Yea i tried to do something with the Interaction enum you mentioned in your first post, but it didn't work out that well. But I like your state idea.

Reply:
Follow up question. Would I add a new field to the State in text_input (I called it "click_state" for now) with the created crate::input::mouse::State? So we got something like this:
`pub struct State {
    is_focused: bool,
    is_pressed: bool,
    is_pasting: Option<Value>,
    cursor: Cursor,
    click_state: crate::input::mouse::State,
}`
with mouse::State(mouse.rs) looking something like this:
`mod button;
mod event;

pub use button::Button;
pub use event::{Event, ScrollDelta};
use std::time::{Duration, SystemTime};

#[derive(Debug, Copy, Clone)]
pub struct State {
    // Maybe we should track which button (left/right) was clicked. don't know if needed
    click_count: usize,
    last_click_position: Option<crate::Point>,
    last_click_timestamp: Option<SystemTime>,
}

// same as before so I'll leave it empty for this comment
impl Default for State {[...]}

impl State {
    // same as before so I'll leave it empty for this comment
    pub fn process_left_click(&mut self, position: crate::Point) -> usize {
        [...]
    }
}`

This feels kinda imperfect right?

Reply:
I would call the field `mouse`.

Then, maybe rename `process_left_click` to `update` and make it:

```rust
pub fn update(&mut self, event: Event) -> Option<Interaction>
```

`Interaction` can be an enum:

```rust
pub enum Interaction {
    Click(Point),
    DoubleClick(Point),
    TripleClick(Point),
}
```

Just brain-storming, this may need a bit more thinking. Feel free to fine-tune it to your liking. I will change any small details before merging.

Reply:
Also, don't feel forced to support more use cases for now. Let's keep it coupled to the `TextInput` needs. We can iterate further in the future.

Reply:
Yeah I keep it coupled to TextInputs needs for now.
I implemented update as `pub fn update(&mut self, position: Point) -> Option<Interaction>` for the next iteration instead of using `event: Event` because I didn't know how else to get the position.